src/main/java/io/datareplication/producer/feed/AssignPagesService.java

```diff
@@ -0,0 +1,158 @@
+        // old (technically-still-current) latest page. If this is same latest page as before, this will make all changes
+        // visible.
+        FeedPageMetadataRepository.PageMetadata newLatestPage;
+        // The previous latest page, if it's different. In that case, this is the final step that will make everything visible.
+        Optional<FeedPageMetadataRepository.PageMetadata> previousLatestPage;
+    }
+
+    Optional<AssignPagesResult> assignPages(
+        Optional<FeedPageMetadataRepository.PageMetadata> maybeLatestPage,
+        List<FeedEntityRepository.PageAssignment> unassignedEntities
+    ) {
+        if (unassignedEntities.isEmpty()) {
+            return Optional.empty();
+        }
+
+        final int generation = maybeLatestPage
+            .map(previousLatestPage -> previousLatestPage.generation() + 1)
+            .orElse(Generations.INITIAL_GENERATION);
+
+        final var assignedEntities = new ArrayList<FeedEntityRepository.PageAssignment>();
+        final var newPages = new ArrayList<FeedPageMetadataRepository.PageMetadata>();
+        var previousLatestPage = Optional.<FeedPageMetadataRepository.PageMetadata>empty();
+
+        // we start with either the old latest page or a new one as the "current" page
+        var currentPage = maybeLatestPage
+            .map(MutablePage::fromLoadedPage)
+            .orElseGet(() -> MutablePage.emptyPage(pageIdProvider.newPageId()));
+        for (var entity : unassignedEntities) {
+            if (!fitsOnPage(currentPage, entity)) {
+                // if this entity doesn't fit on the current page, we close it and create a new one
+                var next = pageIdProvider.newPageId();
+                var finalizedPage = currentPage.finishPage(Optional.of(next), generation);
+                currentPage = MutablePage.emptyPage(next, finalizedPage.pageId());
+
+                if (maybeLatestPage.stream().anyMatch(p -> p.pageId().equals(finalizedPage.pageId()))) {
+                    // if this page has the same ID as the old latest page, we have to store it separately
+                    previousLatestPage = Optional.of(finalizedPage);
+                } else {
+                    // just add it to the pile
+                    newPages.add(finalizedPage);
+                }
+            }
+
+            // update the current page metadata
+            currentPage.append(entity);
+            assignedEntities.add(assignedEntity(entity, currentPage.pageId()));
+        }
+
+        return Optional.of(new AssignPagesResult(
+            assignedEntities,
+            newPages,
+            currentPage.finishPage(Optional.empty(), generation),
+            previousLatestPage
+        ));
+    }
+
+    private boolean fitsOnPage(MutablePage page, FeedEntityRepository.PageAssignment entity) {
+        var pageEmpty = page.numberOfEntities() == 0;
+        var tooManyEntities = page.numberOfEntities() >= maxEntitiesPerPage;
```

Comment:
```suggestion
        var tooManyEntities = page.numberOfEntities() > maxEntitiesPerPage;
```
otherwise the "numberOFEntities" can not be reached (only when it is set to 1)